datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs

```diff
@@ -206,7 +206,16 @@ impl<'a> TreeNodeRewriter for ConstEvaluator<'a> {
 
     fn mutate(&mut self, expr: Expr) -> Result<Expr> {
         match self.can_evaluate.pop() {
-            Some(true) => Ok(Expr::Literal(self.evaluate_to_scalar(expr)?)),
+            Some(true) => {
+                // After simplifying the expression, data_type may change, so we need to cast it.
```

Comment:
This seems wrong to me -- the type of the expression should always be the same before/after simplification. If it isn't the same that is a bug (perhaps we could raise an internal error here?)

Reply:
Correct it